bootc images support

snir911 opened the pull request:
<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->
**- Description of the problem which is fixed/What is the use case**

**- What I did**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

bpradipt commented on config/peerpods/podvm/bootc/usr/lib/bootc/install/00-kargs.toml:
```diff
@@ -0,0 +1,5 @@
+[install.filesystem.root]
+type = "ext4"
+[install]
+kargs = [ "console=ttyS0", "selinux=0", "enforcing=0", "audit=0"]
```
We'll need to enable selinux eventually. May be worth testing without adding the following args `"selinux=0", "enforcing=0", "audit=0"` ?

bpradipt commented on config/peerpods/podvm/lib.sh:
```diff
@@ -424,7 +424,7 @@ function get_podvm_image_format() {
 
     # jq -r when you want to output plain strings without quotes. Otherwise the string will be quoted
-    PODVM_IMAGE_FORMAT=$(qemu-img info -f raw --output json "${image_path}" | jq -r '.format') ||
+    PODVM_IMAGE_FORMAT=$(qemu-img info --output json "${image_path}" | jq -r '.format') ||
```
/lgtm

bpradipt commented on config/peerpods/podvm/aws-podvm-image-handler.sh:
```diff
@@ -317,7 +317,7 @@ function create_ami() {
 
     # Prepare the pause image for embedding into the ami
-    download_and_extract_pause_image "${PAUSE_IMAGE_REPO}" "${PAUSE_IMAGE_VERSION}" "${PAUSE_IMAGE_REPO_AUTH_FILE}"
+    download_and_extract_pause_image "${PAUSE_IMAGE_REPO}" "${PAUSE_IMAGE_VERSION}" "${CLUSTER_PULL_SECRET_AUTH_FILE}"
```
/lgtm
I'm acking individual commits. Looks like some of these commits can be in a separate PR and merged sooner before the bootc one..

bpradipt commented on config/peerpods/podvm/aws-podvm-image-handler.sh:
```diff
@@ -97,7 +97,7 @@ function install_aws_cli() {
 
     # Install aws cli v2
-    unzip -o "/tmp/awscliv2.zip" -d "/tmp" ||
+    unzip -q -o "/tmp/awscliv2.zip" -d "/tmp" ||
```
/lgtm
This can also be clubbed in a separate PR along with first two commits and merged.

bpradipt commented on config/peerpods/podvm/azure-podvm-image-cm.yaml:
```diff
@@ -59,3 +59,18 @@ data:
+
+  # bootc
+  #PODVM_IMAGE_URI: bootc::image-registry.openshift-image-registry.svc:5000/openshift-sandboxed-containers-operator/podvm-bootc
+  # Custom bootc build configuration: https://osbuild.org/docs/bootc/#-build-config
```
Do we want to expose the disk config to user or have a fixed config which is tested and supported ?

snir911 replied:
yes, indeed we can have a default config file

bpradipt commented on config/peerpods/podvm/bootc/README.md:
````diff
@@ -0,0 +1,89 @@
+       "${IMG}"
+```
+Artifact will be located at  output/qcow2/disk.qcow2
+Upload it to your cloud-provider.
````
It might be good to mention that when using qcow2, we can also leverage the `oci` format support in podvm image builder to create a VM image from it

bpradipt commented on config/peerpods/podvm/aws-podvm-image-handler.sh:
```diff
@@ -298,9 +298,6 @@ function create_ami() {
-        # Add Azure yum repositories
-        add_azure_repositories
-
         # Install required rpm packages
```
/lgtm
nit: typo in the commit message - s/insatll/install

beraldoleal commented on config/peerpods/podvm/bootc/README.md:
```diff
@@ -0,0 +1,89 @@
+IMG=quay.io/example/podvm-bootc
+AUTHFILE=/path/to/pull-secret
+podman build --authfile ${AUTHFILE} -f Containerfile.rhel -t ${IMG}
+#podman push ${IMG}
```
nit: remove comment?

snir911 replied:
It's mentioned as an option above

beraldoleal commented on config/peerpods/podvm/bootc/README.md:
````diff
@@ -0,0 +1,89 @@
+registry:
+```
+oc apply -f podvm-git-buildconfig.yaml
+IMG=bootc::image-registry.openshift-image-registry.svc:5000/openshift-sandboxed-containers-operator/podvm-bootc
````
IIUC, this IMG variable is not used for the build process. Or it is?

So maybe it would be nice to move it to the "Convert" section?

snir911 replied:
This is pushed to the registry at this point, if it's not clear maybe i can put this in both sections

beraldoleal replied:
On the first section, IMG is assigned a valid image address. Here, the bootc:: prefix is introduced.

Wouldn't podman pull and podman run fail in this case? Maybe we should define:

```
IMG=image-registry.openshift-image-registry.svc:5000/openshift-sandboxed-containers-operator/podvm-bootc
IMG_URI=bootc::${IMG}
```

And use IMG_URI where is needed?

beraldoleal commented on config/peerpods/podvm/bootc/podvm-git-buildconfig.yaml:
```diff
@@ -0,0 +1,32 @@
+    contextDir: "config/peerpods/podvm/bootc"
+    git:
+      uri: 'https://github.com/openshift/sandboxed-containers-operator.git'
+      ref: devel
```
When reading the current code, I noticed a few references like this, which we might forget to update during the releases. Maybe is time to use a version.yaml file, parse with jq and concentrate the references there?

beraldoleal commented on config/peerpods/podvm/bootc/Containerfile.rhel:
```diff
@@ -66,12 +66,12 @@ RUN export OS_VERSION_MAJOR=$(grep "^VERSION=" /etc/os-release | cut -d '=' -f 2
+
+ADD nvidia/nvidia-cdi.service /etc/systemd/system/nvidia-cdi.service
+ADD nvidia/generate-nvidia-cdi.sh /usr/local/bin/generate-nvidia-cdi.sh
+RUN ln -s /etc/systemd/system/nvidia-cdi.service /etc/systemd/system/multi-user.target.wants/nvidia-cdi.service
```
It would be nice to squash this commit with the nvidia one.

beraldoleal reviewed (commented):
Hi @snir911, thanks for this! Overall, it looks good, I just left a few comments.

I have a better sense now of what needs to be implemented to use bootc with GCP. I started exploring this here (using mkosi):

https://github.com/openshift/sandboxed-containers-operator/pull/519

Feel free to comment there if you anticipate any issues.

That said, what’s the plan for transitioning to bootc? I see this is being added as an additional method, but how do we intend to make it the default?

snir911 commented:
This is on top of #543 now and includes it

beraldoleal commented:
@snir911 , now that #543 was merged, do you mind to rebase this one? And maybe we can remove the `do-not-merge` flag? Or are you still waiting for something?

snir911 commented on config/peerpods/podvm/bootc/Containerfile.rhel:
```diff
@@ -0,0 +1,82 @@
+
+# afterburn is required for Azure
+#RUN dnf config-manager --add-repo=https://mirror.stream.centos.org/9-stream/AppStream/x86_64/os/ && dnf install -y --nogpgcheck afterburn && dnf clean all && dnf config-manager --set-disabled "*centos*"
+RUN dnf install -y afterburn && dnf clean all
```
Installing afterburn is the only thing requires a subscription and AFAIK, required only by Azure.
Should we drop it? this will allow building bootc image in-cluster using the buildconfig definition without any special configuration.
@beraldoleal do you know if afterburn (or any other pkg) is required for gcp?

beraldoleal replied:
@snir911 I haven’t used it directly during development, so my initial assumption is that it’s not needed and can be dropped. If it was used, it must have been under the hood, and I didn’t notice.

That said, we can revisit this if we run into any issues with GCP later.

bpradipt replied:
Afaik afterburn is used specifically for azure in mkosi build today to indicate that the VM is ready.
https://github.com/confidential-containers/cloud-api-adaptor/commit/a70e147768d8026f4fe67ea45008355c165408f5

wainersm replied:
@snir911 what if have an argument (`CLOUD_PROVIDER`) that if "azure" then subscribe && install afterburn?

snir911 commented on config/peerpods/podvm/bootc/Containerfile.rhel:
```diff
@@ -0,0 +1,82 @@
+RUN umoci init --layout pause && umoci new --image pause:k8s
+RUN umoci config --image pause:k8s --config.entrypoint=/pause --author="OSC"
+RUN umoci unpack --rootless --image pause:k8s /pause
+COPY --from=pause-bin /pause /pause/rootfs/pause
```
@bpradipt do you have more elegant suggestion for creating the pause bundle? (better to avoid binary download)

bpradipt replied:
Which binary are you referring to ?

snir911 replied:
I wasn't clear, I'd prefer to avoid using umoci which doesn't seems to exist in normal repos

bpradipt replied:
Hmm, I don't know if there is any alternative to umoci for unpacking an OCI image which can then be used as-is by a container runtime.  May be `podman run` followed by a `podman export` could help but not sure..

snir911 commented on config/peerpods/podvm/bootc/nvidia/generate-nvidia-cdi.sh:
```diff
@@ -0,0 +1,10 @@
+#nvidia-persistenced
+# set confidential compute to ready state
+#nvidia-smi conf-compute -srs 1 
+# Generate NVIDIA CDI configuration
```
the commented are required for CoCo, i'm not sure how to differentiate

wainersm replied:
maybe have the `CONFIDENTIAL_COMPUTE_ENABLED` passed as argument here?

snir911 replied:
This is a service running inside the VM I'm not sure how to do it so it will be taken into account at runtime

beraldoleal commented on scripts/ami-helper/ami-helper.sh:
```diff
@@ -0,0 +1,193 @@
+	oc get ns openshift-sandboxed-containers-operator >/dev/null 2>&1 || (echo "OSC namespace is missing, re-run after installting the operator" && exit 1)
+
+	cat <<EOF > "${CREDENTIALS_REQUEST_YAML_FILE}"
+apiVersion: cloudcredential.openshift.io/v1
```
I noticed some variables here, like ${BUCKET_NAME}, and that might be why you're hardcoding it. But if it's not too much work, I suggest consolidating all requests under `config/peerpods/credentials-requests`. In this case, some substitution would be needed, but I still think having them there would help with our organization.

snir911 replied:
Discussed offline, i set AWS credentials for image creation in separate until this will be the default method

beraldoleal commented on config/peerpods/podvm/aws-podvm-image-handler.sh:
```diff
@@ -318,6 +439,33 @@ function create_ami() {
+    # If any error occurs, exit the script with an error message
+
+    # Install packages if INSTALL_PACKAGES is set to yes
+    if [[ "${INSTALL_PACKAGES}" == "yes" ]]; then
```
since you are adding this installation steps here (which makes sense) I believe we have it duplicated under creater_ami_from_scratch() method. IIUC, when following the from scratch flow, it would be installing twice. Should we remove there? Or am I missing something?

beraldoleal commented on config/peerpods/podvm/aws-podvm-image-handler.sh:
```diff
@@ -287,10 +312,106 @@ function delete_ami_id_annotation_from_peer_pods_cm() {
+        # Form the path of the podvm vhd image.
+        podvm_image_path="${extraction_destination_path}/rootfs/${PODVM_IMAGE_SRC_PATH}"
+
+        upload_image_to_s3
```
Maybe I'm missing something here too. But with Azure, you are uploading the disk and creating image outside the bootc) and oci) case. IIUC, for the bootc case here, the image would not be uploaded.

Did I got this wrong?

beraldoleal replied:
I think I got it... Azure uses `bootc_to_qcow2()`, while AWS uses `bootc_to_ami()`. They seem to do similar things, but `bootc_to_ami()` also handles the upload, while `bootc_to_qcow2()` doesn’t.

Please, consider keeping the upload step separate in AWS too, like it’s done in Azure... That way, both follow the same flow, and it’s easier to keep things consistent

snir911 replied:
bootc image builder supports native upload for AWS, hence i used it, other cloud providers are expected to be supported in the future as well.
I can implement the upload myself for AWS, however, i believe it will increase the points of failure

beraldoleal replied:
I think splitting it would help keep things consistent, but I'll leave that decision for this PR up to you, we might split it later.

Said that, if we keep it as-is, the function name could be clearer since `bootc_to_ami` sounds like it just creates the AMI.

wainersm replied:
@snir911 ,

don't you need to convert qcow -> raw before `upload_image_to_s3`?

bpradipt replied:
Looks like bootc-image-builder either allows you to create qcow2/raw or directly create AMI. There is no intermediate step of using a qcow2 or raw and create ami. At least I couldn't find it in the documents.

wainersm replied:
The block where `upload_image_to_s3` is called will handle the oci case.

wainersm commented on config/peerpods/podvm/bootc/Containerfile.rhel:
```diff
@@ -0,0 +1,82 @@
+# build pause
+FROM registry.k8s.io/pause:3.9 as pause-bin
```
Hi @snir911 ,

An user building a custom operator's built podvm can customize the pause image by setting the `PAUSE_IMAGE_REPO`, `PAUSE_IMAGE_VERSION` and `CLUSTER_PULL_SECRET_AUTH_FILE` variable. Would make sense to pass these as build arguments in this containerfile too?

Apart from that, maybe we want to keep the pause image version in sync with the operator's built podvm, i.e,  https://github.com/openshift/sandboxed-containers-operator/blob/devel/config/peerpods/podvm/lib.sh#L8

Example:
```
ARG PAUSE_IMAGE_REPO="quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256"
ARG PAUSE_IMAGE_VERSION="7f3cb6f9d265291b47a7491c2ba4f4dd0752a18b661eee40584f9a5dbcbe13bb"

FROM ${PAUSE_IMAGE_REPO}:${PAUSE_IMAGE_VERSION} as pause-bin
```

snir911 replied:
OCP's pause image is much more heavy so i preferred the k8s one (before we couldn't use that)

about the AUTH_FILE this is indeed needed, I will address in separate

snir911 replied:
I think this won't be needed as the global authfile is passed and soon we'll have also support in imagePullSecret for custom secrets

bpradipt replied:
The imagePullSecret is not for the pause image. Note that the secret that we use here is for downloading and embedding the pause image.

wainersm commented on config/peerpods/podvm/bootc/Containerfile.rhel:
```diff
@@ -0,0 +1,82 @@
+RUN tar -xzvf podvm-binaries.tar.gz -C /
+RUN sed -i 's#What=/kata-containers#What=/var/kata-containers#g' /etc/systemd/system/run-kata\\x2dcontainers.mount
+# until created in the payload:
+RUN ln -s /run/peerpod/policy.rego /etc/kata-opa/default-policy.rego
```
We might need a more complex handler like did in https://github.com/openshift/sandboxed-containers-operator/pull/553

snir911 replied:
this can be removed now is it's included in the payload

wainersm commented on config/peerpods/podvm/bootc/Containerfile.rhel:
```diff
@@ -39,6 +39,42 @@ COPY --from=payload /podvm-binaries.tar.gz /podvm-binaries.tar.gz
 
+
+
+########## Nvidia podVM target ##########
```
operator's built podvm nvidia gpu is optional and protected by `ENABLE_NVIDIA_GPU=no` by default. Wouldn't make sense to have it optional here too?

snir911 replied:
It's optional, it won't built it unless target nvidia-podvm-bootc is explicitly set.
BTW I'll drop all the unneeded Nvidia flags when bootc will become the default

wainersm replied:
ah, got it. Makes sense.

wainersm commented on config/peerpods/podvm/lib.sh:
```diff
@@ -457,6 +461,89 @@ function validate_podvm_image() {
+        echo "Using Custom Bootc Build Configuration"
+        echo "${BOOTC_BUILD_CONFIG}" >> ./config.toml
+    else
+        echo "Using Default Bootc Build Configuration"
```
Why don't you use the [config.toml](https://github.com/openshift/sandboxed-containers-operator/pull/484/commits/d6b6cceec908fc95e7509203f200bb42cfd89434#diff-a8633a2cbbc3a90771c8cbd93128222dd9007c8f97989b5fcd8429e8e4958e81) versioned on this repo?

I would also rename it to `bootc-config.toml` as `config.toml` is pretty generic and it might end up overwriting a file you don't want to (and introducing a silent bug).

wainersm commented on config/peerpods/podvm/bootc/Containerfile.rhel:
```diff
@@ -0,0 +1,44 @@
+    subscription-manager register --org=${ORG_ID} --activationkey=${ACTIVATION_KEY}; \
+    fi
+
+COPY etc /etc
```
Suggestion: move the files that you will copy to the bootc image to under `files`. Like this:

```
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	renamed:    etc/sudoers.d/wheel-nopasswd -> files/etc/sudoers.d/wheel-nopasswd
	renamed:    etc/systemd/system/afterburn-checkin.service -> files/etc/systemd/system/afterburn-checkin.service
	renamed:    usr/lib/bootc/install/00-kargs.toml -> files/usr/lib/bootc/install/00-kargs.toml
```

snir911 replied:
I can do that, i saw this is commonly done this way in the bootc examples

wainersm replied:
We can keep it as is for now.

snir911 committed:
podvm: add basic logic to support bootc images & pre-built disks

supports URIs with the following pattern:
bootc::quay.io/openshift_sandboxed_containers/azure-bootc-podvm-image:latest
or for pre-built disk artifact
oci::quay.io/osc/disks:podvm:/path/to/qcow2

bootc container image build:
$ podman build --authfile <auth-file> --build-arg ORG_ID=<org-id> --build-arg ACTIVATION_KEY=<key>
 -f Containerfile.rhel -t podvm-bootc
create VM disk:
$ mkdir output
$ sudo podman run \
       -it --rm \
       --privileged \
       --security-opt label=type:unconfined_t \
       --authfile ${AUTHFILE} \
       -v $(pwd)/config.toml:/config.toml:ro \
       -v $(pwd)/output:/output \
       -v /var/lib/containers/storage:/var/lib/containers/storage \
       registry.redhat.io/rhel9/bootc-image-builder:latest \
       --type qcow2 \
       --rootfs xfs \
       --local \
       podvm-bootc

Signed-off-by: Snir Sheriber <ssheribe@redhat.com>

snir911 committed:
podvm: azure: add bootc images support

and also support precreated disk in oci container

build bootc podvm container image:
$ podman build --authfile <auth-file> \
  --build-arg CLOUD_PROVIDER=azure --build-arg ORG_ID=<org-id> \
  --build-arg ACTIVATION_KEY=<key> -f Containerfile.rhel -t podvm-bootc

Signed-off-by: Snir Sheriber <ssheribe@redhat.com>

snir911 committed:
podvm: aws: add bootc image support

and pre-built disks support

Signed-off-by: Snir Sheriber <ssheribe@redhat.com>

snir911 committed:
podvm: aws: add helper script to support

bucket and permissions creation for bootc and disks
upload

Signed-off-by: Snir Sheriber <ssheribe@redhat.com>

snir911 committed:
podvm: bootc usage documention

Signed-off-by: Snir Sheriber <ssheribe@redhat.com>

snir911 committed:
podvm: add build config for bootc

in-cluster builds

Signed-off-by: Snir Sheriber <ssheribe@redhat.com>

snir911 commented:
address most of the comments and split the the commits for easier reading

beraldoleal commented on config/peerpods/podvm/bootc/Containerfile.rhel:
```diff
@@ -0,0 +1,39 @@
+COPY --from=pause-bin /pause /pause/rootfs/pause
+
+# Get payload
+FROM registry.redhat.io/openshift-sandboxed-containers/osc-podvm-payload-rhel9:1.8.0-6 as payload
```
@snir911,

I'm trying to understand how/when we are going to update this to 1.9.0.

IIUC, after this gets merged and included in our next build, deployments of that build would be using an old image without the recent changes. Maybe we need this to be parameterized?

beraldoleal commented on config/peerpods/podvm/bootc/README.md:
````diff
@@ -0,0 +1,114 @@
+  [[customizations.filesystem]]
+  mountpoint = "/var/kata-containers"
+  minsize = "15 GiB"
+```
````
Hi @snir911, 

I'm thinking in terms of steps the users needs  to configure/deploy this. 

If I got this right, for AWS and Azure (when using the "from scratch" approach), there's no need for provider-podvm-image-cm, and things just work automatically.

But if we switch to the prebuilt method (including for GCP), this ConfigMap becomes mandatory to point to the image. So in the future, if prebuilt becomes the default, could we drop this -cm, and concentrate all settings at `peer-pods-cm`? Or would this additional cm still be needed to set the image uri?

snir911 replied:
I think in any case it's preferred to have special settings for image creation and a separate one for runtime (peer-pod-cm) which is heavily attached to upstream.
( BTW we could simply set by default a uri in the CM)

beraldoleal replied:
I just would like to see less cm needed to bootstrap this, to reduce the installation time. Default values that works out of the box are fine. As long users don't need to deploy new cm.

beraldoleal commented on config/peerpods/podvm/bootc/Containerfile.rhel:
```diff
@@ -9,7 +9,7 @@ RUN umoci unpack --rootless --image pause:k8s /pause
 
 # Get payload
-FROM registry.redhat.io/openshift-sandboxed-containers/osc-podvm-payload-rhel9:1.8.0-6 as payload
+FROM registry.redhat.io/openshift-sandboxed-containers/osc-podvm-payload-rhel9:1.9.0 as payload
```
This change should be squashed with previous commit that was introducing this. In any case, this is not a valid reference today, right?

beraldoleal commented on config/peerpods/podvm/aws-podvm-image-handler.sh:
```diff
@@ -362,8 +362,19 @@ function upload_image_to_s3() {
     # Check if the image exists
     [[ ! -f "${podvm_image_path}" ]] && error_exit "Image does not exist"
 
+    get_podvm_image_format ${podvm_image_path}
```
maybe squash this with 03a1095d39cf6f9a9e80ce33153086c372af2fd1

beraldoleal commented on config/peerpods/podvm/bootc/podvm-git-buildconfig.yaml:
```diff
@@ -0,0 +1,32 @@
+    contextDir: "config/peerpods/podvm/bootc"
+    git:
+      uri: 'https://github.com/openshift/sandboxed-containers-operator.git'
+      ref: devel
```
Reading again, I'm concerned of something else... if this is always pointing to devel, that means that anyone in production using podvm images from bootc would be using code from devel branch? If so, maybe we can add a `# TODO-RELEASE: ` comment on top of it, so we can easly grep and replace values before creating the release branch?

snir911 replied:
would it make sense to switch ref to release-1.9? (not exist yet but still..)

beraldoleal replied:
I think that as long the code exists on devel branch, it should be devel. But we need a strategy to when we are creating the "release branch". That is the reason I mentioned the # TODO-RELEASE. But I think @gkurz and @c3d had used `## OSC_VERSION` before. not sure how that tag is used and if have the same goal, tho.

beraldoleal reviewed (commented):
Hi @snir911, thanks again for pushing this! I think we're really close. I'm eager to get this merged ASAP, but I left a few minor comments/suggestions for you. :)

bpradipt commented on config/peerpods/podvm/bootc/README.md:
```diff
@@ -0,0 +1,114 @@
+1. bake an OCI container image with you disk file
+2. push it to some container registry
+3. set in the `<cloud-provider>-podvm-image-cm` IMAGE_TYPE: pre-built and PODVM_IMAGE_URI
+that follows this format: `oci::<conatiner-image-uri>::</absolute/path/to/disk>`
```
s/conatiner-image-uri/container-image-uri

bpradipt commented on config/peerpods/podvm/Dockerfile.podvm-builder:
```diff
@@ -26,6 +26,9 @@ RUN mkdir -p /scripts
 
 ADD lib.sh libvirt-podvm-image-handler.sh aws-podvm-image-handler.sh azure-podvm-image-handler.sh /scripts/
 
+RUN mkdir -p /scripts/bootc
```
nit, typo in the commit message
s/precarted/pre-created 
or s/precarted/pre-built

wainersm replied:
While in the commit message, `--authfile ${AUTHFILE} --authfile <auth-file>` passing --authfile twice.

wainersm commented on config/peerpods/podvm/bootc/Containerfile.rhel:
```diff
@@ -42,6 +42,42 @@ COPY --from=payload /podvm-binaries.tar.gz /podvm-binaries.tar.gz
+    export DRIVER_STREAM=$(echo ${DRIVER_VERSION} | cut -d '.' -f 1) && \
+    dnf config-manager --add-repo=https://developer.download.nvidia.com/compute/cuda/repos/rhel9/x86_64/cuda-rhel9.repo && \
+    dnf config-manager --best --nodocs --setopt=install_weak_deps=False --save && \
+    dnf -y module enable nvidia-driver:${DRIVER_STREAM}/default && \
```
Hi @snir911 !

Tried to build it locally, it failed to install nvidia-driver package as you can see below. Do I need to pass the activation key for nvidia case too?


```
wmoschet@wmoschet-thinkpadp16vgen1:~/src/github.com/openshift/sandboxed-containers-operator/config/peerpods/podvm/bootc$ sudo podman build --authfile ~/Documents/credentials/pull-secret-brew -f Containerfile.rhel --target nvidia-podvm-bootc -t nvidia-podvm-bootc
WARN[0000] missing "ORG_ID" build argument. Try adding "--build-arg ORG_ID=<VALUE>" to the command line 
WARN[0000] missing "ACTIVATION_KEY" build argument. Try adding "--build-arg ACTIVATION_KEY=<VALUE>" to the command line 
WARN[0000] missing "CLOUD_PROVIDER" build argument. Try adding "--build-arg CLOUD_PROVIDER=<VALUE>" to the command line 

<snip>

[5/5] STEP 1/10: FROM 9ed3e93619b1b985c74aa52d367bd68695d58af52584a0d3181f93de994f1234 AS nvidia-podvm-bootc
[5/5] STEP 2/10: ENV DRIVER_VERSION="550.144.03"
--> 12b75f53a954
[5/5] STEP 3/10: RUN export OS_VERSION_MAJOR=$(grep "^VERSION=" /etc/os-release | cut -d '=' -f 2 | sed 's/"//g' | cut -d '.' -f 1) &&     export KERNEL_VERSION=$(rpm -q --qf "%{VERSION}" kernel-core) &&     export KERNEL_RELEASE=$(rpm -q --qf "%{RELEASE}" kernel-core | sed 's/\.el.\(_.\)*$//') &&     export DRIVER_STREAM=$(echo ${DRIVER_VERSION} | cut -d '.' -f 1) &&     dnf config-manager --add-repo=https://developer.download.nvidia.com/compute/cuda/repos/rhel9/x86_64/cuda-rhel9.repo &&     dnf config-manager --best --nodocs --setopt=install_weak_deps=False --save &&     dnf -y module enable nvidia-driver:${DRIVER_STREAM}/default &&     dnf install -y kmod-nvidia-${DRIVER_VERSION}-${KERNEL_VERSION}-${KERNEL_RELEASE}         nvidia-driver-${DRIVER_VERSION}         nvidia-driver-cuda-${DRIVER_VERSION}         nvidia-driver-libs-${DRIVER_VERSION}         nvidia-persistenced-${DRIVER_VERSION}         nvidia-driver-NVML-${DRIVER_VERSION}         nvidia-container-toolkit &&     dnf clean all
Updating Subscription Management repositories.
Unable to read consumer identity

This system is not registered with an entitlement server. You can use subscription-manager to register.

Adding repo from: https://developer.download.nvidia.com/compute/cuda/repos/rhel9/x86_64/cuda-rhel9.repo
Updating Subscription Management repositories.
Unable to read consumer identity

This system is not registered with an entitlement server. You can use subscription-manager to register.

Updating Subscription Management repositories.
Unable to read consumer identity

This system is not registered with an entitlement server. You can use subscription-manager to register.

cuda-rhel9-x86_64                               898 kB/s | 2.5 MB     00:02    
Last metadata expiration check: 0:00:01 ago on Thu Mar  6 14:22:24 2025.
Ignoring unnecessary profile: 'nvidia-driver/default'
Dependencies resolved.
================================================================================
 Package           Architecture     Version             Repository         Size
================================================================================
Enabling module streams:
 nvidia-driver                      550                                        

Transaction Summary
================================================================================

Complete!
Updating Subscription Management repositories.
Unable to read consumer identity

This system is not registered with an entitlement server. You can use subscription-manager to register.

Last metadata expiration check: 0:00:02 ago on Thu Mar  6 14:22:24 2025.
Error: 
 Problem 1: conflicting requests
  - nothing provides xorg-x11-server-Xorg(x86-64) >= 1.19.0-3 needed by nvidia-driver-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
 Problem 2: conflicting requests
  - nothing provides ocl-icd needed by nvidia-driver-cuda-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
  - nothing provides opencl-filesystem needed by nvidia-driver-cuda-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
 Problem 3: cannot install the best candidate for the job
  - nothing provides libX11.so.6()(64bit) needed by nvidia-driver-libs-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
  - nothing provides libXext.so.6()(64bit) needed by nvidia-driver-libs-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
  - nothing provides vulkan-loader needed by nvidia-driver-libs-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
  - nothing provides libdrm.so.2()(64bit) needed by nvidia-driver-libs-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
  - nothing provides libgbm.so.1()(64bit) needed by nvidia-driver-libs-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
  - nothing provides libglvnd(x86-64) >= 1.0 needed by nvidia-driver-libs-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
  - nothing provides libglvnd-egl(x86-64) >= 1.0 needed by nvidia-driver-libs-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
  - nothing provides libglvnd-gles(x86-64) >= 1.0 needed by nvidia-driver-libs-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
  - nothing provides libglvnd-glx(x86-64) >= 1.0 needed by nvidia-driver-libs-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
  - nothing provides libglvnd-opengl(x86-64) >= 1.0 needed by nvidia-driver-libs-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
  - nothing provides libvdpau(x86-64) >= 0.5 needed by nvidia-driver-libs-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
 Problem 4: package nvidia-persistenced-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64 requires nvidia-driver-cuda = 3:550.144.03, but none of the providers can be installed
  - conflicting requests
  - nothing provides ocl-icd needed by nvidia-driver-cuda-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
  - nothing provides opencl-filesystem needed by nvidia-driver-cuda-3:550.144.03-1.el9.x86_64 from cuda-rhel9-x86_64
(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)
Error: building at STEP "RUN export OS_VERSION_MAJOR=$(grep "^VERSION=" /etc/os-release | cut -d '=' -f 2 | sed 's/"//g' | cut -d '.' -f 1) &&     export KERNEL_VERSION=$(rpm -q --qf "%{VERSION}" kernel-core) &&     export KERNEL_RELEASE=$(rpm -q --qf "%{RELEASE}" kernel-core | sed 's/\.el.\(_.\)*$//') &&     export DRIVER_STREAM=$(echo ${DRIVER_VERSION} | cut -d '.' -f 1) &&     dnf config-manager --add-repo=https://developer.download.nvidia.com/compute/cuda/repos/rhel9/x86_64/cuda-rhel9.repo &&     dnf config-manager --best --nodocs --setopt=install_weak_deps=False --save &&     dnf -y module enable nvidia-driver:${DRIVER_STREAM}/default &&     dnf install -y kmod-nvidia-${DRIVER_VERSION}-${KERNEL_VERSION}-${KERNEL_RELEASE}         nvidia-driver-${DRIVER_VERSION}         nvidia-driver-cuda-${DRIVER_VERSION}         nvidia-driver-libs-${DRIVER_VERSION}         nvidia-persistenced-${DRIVER_VERSION}         nvidia-driver-NVML-${DRIVER_VERSION}         nvidia-container-toolkit &&     dnf clean all": while running runtime: exit status 1
```

snir911 replied:
that's odd, works for me, maybe cache issue?

snir911 replied:
have you ran it with subscription key?

wainersm commented on config/peerpods/podvm/bootc/Containerfile.rhel:
```diff
@@ -0,0 +1,39 @@
+
+# register
+RUN if [[ -n "${ACTIVATION_KEY}" && -n "${ORG_ID}" ]]; then \
+    #rm -f /etc/rhsm-host && rm -f /etc/pki/entitlement-host; \
```
I didn't test with the activation path, should this line be enabled (or removed)?

snir911 replied:
I saw it was mentioned somewhere it's needed

wainersm replied:
Hi @snir911 , the line is commented "#"

wainersm commented on config/peerpods/podvm/lib.sh:
```diff
@@ -459,6 +463,82 @@ function validate_podvm_image() {
+# 3. auth_json_file (can be empty): Path to the registry secret file to use for downloading the image.
+# 4. run_args: arguments for the podman run command that runs bootc image builder
+# 5. bib_args: arguments for bootc image builder itself
+# Output: based on input, either, qcow2 image at output/qcow2/disk.qcow2 or ami
```
Nit: don't generate an ami

snir911 replied:
why not? it will for aws

wainersm replied:
Never mind...

wainersm commented on config/peerpods/podvm/lib.sh:
```diff
@@ -459,6 +463,82 @@ function validate_podvm_image() {
+    # some VM customizations
+    if [[ -n "${BOOTC_BUILD_CONFIG}" ]]; then
+        echo "Using Custom Bootc Build Configuration"
+        echo "${BOOTC_BUILD_CONFIG}" >> ./custom-config.toml
```
Here you want to copy the content of `${BOOTC_BUILD_CONFIG}`  to `./custom-config.toml`, so either `cp -f "${BOOTC_BUILD_CONFIG}" ./custom-config.toml` or  `cat "${BOOTC_BUILD_CONFIG}" > ./custom-config.toml`

wainersm replied:
Ignore this comment pls. I set `BOOTC_BUILD_CONFIG` as the path to the config, but it should be the file's content....

wainersm reviewed (approved):
Hi @snir911 ,

Thanks for gracefully addressing my comments and being calm ("Keep Calm and Ignore this wainer's comment") with my nonsense comments.

LGTM

gkurz commented on config/peerpods/podvm/bootc/podvm-git-buildconfig.yaml:
```diff
@@ -0,0 +1,33 @@
+  source:
+    contextDir: "config/peerpods/podvm/bootc"
+    git:
+      uri: 'https://github.com/openshift/sandboxed-containers-operator.git'
```
Am I understanding correctly that this is going to access the upstream git repo ? If so, is there a chance this can ever work in disconnected setups ?

snir911 commented:
Yes, this is not intended to be used in production, ideally we will either
ship a pre built bootc image or pre built disk in OCI container

On Fri, 7 Mar 2025, 18:26 Greg Kurz, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In config/peerpods/podvm/bootc/podvm-git-buildconfig.yaml
> <https://github.com/openshift/sandboxed-containers-operator/pull/484#discussion_r1985357033>
> :
>
> > +    local: false
> +---
> +apiVersion: build.openshift.io/v1
> +kind: BuildConfig
> +metadata:
> +  name: bootc-image-build
> +  namespace: openshift-sandboxed-containers-operator
> +spec:
> +  output:
> +    to:
> +      kind: ImageStreamTag
> +      name: podvm-bootc:latest
> +  source:
> +    contextDir: "config/peerpods/podvm/bootc"
> +    git:
> +      uri: 'https://github.com/openshift/sandboxed-containers-operator.git'
>
> Am I understanding correctly that this is going to access the upstream git
> repo ? If so, is there a chance this can ever work in disconnected setups ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/sandboxed-containers-operator/pull/484#pullrequestreview-2667767655>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADAS4IXUINYVHGS4QXDZSE32THCEDAVCNFSM6AAAAABS3AJ7MGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMNRXG43DONRVGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> .com>
>

wainersm commented:
I noticed @snir911 dropped Nvidia support on last update. I'm okay with that, my ack stands.

beraldoleal reviewed (approved):
LGTM, thanks.

beraldoleal commented:
Hi @gkurz can we merge this? Are your questions answered?

beraldoleal commented:
@snir911 just told me, he might have found an minor issue (not critical to the bootc workflow). He is working on an issue.

gkurz commented:
> Hi @gkurz can we merge this? Are your questions answered?

I did not have time to review as I would have liked to but I won't block the merge.

beraldoleal commented:
Thanks @snir911 .
